View purity report

CupcakeFoxxo opened the pull request.

Commit:
Added View Purity Report button

Commit:
Added skeleton code to view purity reports

ahirschberg committed:
Refactor reports to use inheritance

Note: report details screen no longer works, needs refactor to general
report details screen (right now it assumes a SourceReport)

Commit:
Merge branch 'issues47/report-refactor' into ViewPurityReport

# Conflicts:
#	src/controller/ViewReportsScreenController.java

Commit:
Can now view purity reports and availability reports separately

Commit:
Added permissions check for viewing puriy reports

ahirschberg commented on src/controller/ViewReportsScreenController.java:
```diff
@@ -34,7 +43,12 @@ public void registerMainApp(MainFXApplication main) {
      * Updates the reports list with stringified reports.
      */
     public void setReportsList() {
```
I think we should pass a `Stream<? extends Report>` into this method, then vary the type of report using that.

ahirschberg commented on src/controller/ViewReportsScreenController.java:
```diff
@@ -34,7 +43,12 @@ public void registerMainApp(MainFXApplication main) {
-        Stream<Report> reports = main.getReportManager().getSourceReports();
+        Stream<? extends Report> reports;
+        if (viewPurityReports) {
+            reports = main.getReportManager().getPurityReports();
```
This is a violation of the 'tell don't ask' principle in OOP. Instead of asking `main` for the report manager and then the purity reports list from the report manager, we should just pass a reference to the reports we want to display into  the controller. That lets us easily reuse this class for any subclass of report.

ahirschberg commented on src/fxapp/MainFXApplication.java:
```diff
@@ -105,10 +107,17 @@ public void setSourceReportScene() {
     
+    public void setViewPurityScene() {
+        viewReports.showPurityReports(true);
+        viewReports.setReportsList();
```
`viewReports.setReportsList()` should get passed the stream of reports, IMO

ahirschberg reviewed (changes requested):
Functionality looks good. As far as the code design goes, I think we should have the ViewReportsScreenController getting passed a stream. The way we have it right now is hard coding which type of report is being shown based on a boolean. This is fragile because: what if we want to add additional report types to display?

That ties into the dependency injection idea he's been talking about in class, where we are injecting the report type (via the stream) into the class dynamically.

Commit:
Changed program to robustly adhere to solid agile design principles for reusable software based on static inheretance using immutable object oriented programming.
[also viewreportscontroller is now passed reports for it to display]

ahirschberg committed:
Fix bugs with my initial branch (sorry)

Fix minor code style.

Since we don't yet need the map pins for the purity reports to be
clickable, it currently prints an error message

jeremydiba committed:
fixed purity report tostring
